### ISSUE TITLE
NullPointerException during cascading saves of bidirectional domain objects

### ISSUE THREAD

**Issue by joemccall86** [grails/grails-core#9290]
Reproducible example: https://github.com/joemccall86/grails-bidirectional-cascade-test (includes working 2.4.5 project and non-working 3.0.8 project).

This is behavior that worked in Grails 2.4.5, but seems broken in Grails 3.0.x. Given a domain class `Team` that `hasMany = [players: Player]`, and each player `hasOne = [contract: Contract]`, and the bi-directional associations are setup properly, a `save(flush: true)` on a `Team` object should cascade saves down to `Player` and `Contract`. What happens instead is a `NullPointerException`. 

This seems very similar to https://github.com/grails/grails-core/issues/698. Running JDK 8 on Windows 10.

Here is the stacktrace:

```
java.lang.NullPointerException
    at org.hibernate.engine.spi.BatchFetchQueue.removeBatchLoadableEntityKey(BatchFetchQueue.java:163)
    at org.hibernate.engine.internal.StatefulPersistenceContext.addEntity(StatefulPersistenceContext.java:389)
    at org.hibernate.engine.internal.StatefulPersistenceContext.addEntity(StatefulPersistenceContext.java:462)
    at org.hibernate.action.internal.AbstractEntityInsertAction.makeEntityManaged(AbstractEntityInsertAction.java:143)
    at org.hibernate.engine.spi.ActionQueue.addResolvedEntityInsertAction(ActionQueue.java:203)
    at org.hibernate.engine.spi.ActionQueue.addInsertAction(ActionQueue.java:181)
    at org.hibernate.engine.spi.ActionQueue.addAction(ActionQueue.java:216)
    at org.hibernate.event.internal.AbstractSaveEventListener.addInsertAction(AbstractSaveEventListener.java:334)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:289)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:195)
    at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:126)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:209)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:194)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.performSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:114)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:90)
    at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onSaveOrUpdate(ClosureEventTriggeringInterceptor.java:113)
    at org.hibernate.internal.SessionImpl.fireSaveOrUpdate(SessionImpl.java:684)
    at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:676)
    at org.hibernate.engine.spi.CascadingActions$5.cascade(CascadingActions.java:235)
    at org.hibernate.engine.internal.Cascade.cascadeToOne(Cascade.java:350)
    at org.hibernate.engine.internal.Cascade.cascadeAssociation(Cascade.java:293)
    at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:161)
    at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:118)
    at org.hibernate.event.internal.AbstractSaveEventListener.cascadeAfterSave(AbstractSaveEventListener.java:470)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:295)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:195)
    at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:126)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:209)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:194)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.performSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:114)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:90)
    at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onSaveOrUpdate(ClosureEventTriggeringInterceptor.java:113)
    at org.hibernate.internal.SessionImpl.fireSaveOrUpdate(SessionImpl.java:684)
    at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:676)
    at org.hibernate.engine.spi.CascadingActions$5.cascade(CascadingActions.java:235)
    at org.hibernate.engine.internal.Cascade.cascadeToOne(Cascade.java:350)
    at org.hibernate.engine.internal.Cascade.cascadeAssociation(Cascade.java:293)
    at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:161)
    at org.hibernate.engine.internal.Cascade.cascadeCollectionElements(Cascade.java:379)
    at org.hibernate.engine.internal.Cascade.cascadeCollection(Cascade.java:319)
    at org.hibernate.engine.internal.Cascade.cascadeAssociation(Cascade.java:296)
    at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:161)
    at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:118)
    at org.hibernate.event.internal.AbstractSaveEventListener.cascadeAfterSave(AbstractSaveEventListener.java:470)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:295)
    at org.hibernate.event.internal.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:195)
    at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:126)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:209)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:194)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.performSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:114)
    at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:90)
    at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onSaveOrUpdate(ClosureEventTriggeringInterceptor.java:113)
    at org.hibernate.internal.SessionImpl.fireSaveOrUpdate(SessionImpl.java:684)
    at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:676)
    at org.hibernate.internal.SessionImpl.saveOrUpdate(SessionImpl.java:671)
    at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave_closure3(AbstractHibernateGormInstanceApi.groovy:255)
    at groovy.lang.Closure.call(Closure.java:426)
    at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:198)
    at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:142)
    at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:112)
    at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave(AbstractHibernateGormInstanceApi.groovy:254)
    at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.save(AbstractHibernateGormInstanceApi.groovy:179)
    at org.grails.datastore.gorm.GormEntity$Trait$Helper.save(GormEntity.groovy:165)
    at baseball.test.BiDirectionalAssociationSpecSpec.$tt__setupData(BiDirectionalAssociationSpecSpec.groovy:42)
    at baseball.test.BiDirectionalAssociationSpecSpec.setupData_closure1(BiDirectionalAssociationSpecSpec.groovy)
    at groovy.lang.Closure.call(Closure.java:426)
    at groovy.lang.Closure.call(Closure.java:442)
    at grails.transaction.GrailsTransactionTemplate$1.doInTransaction(GrailsTransactionTemplate.groovy:67)
    at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
    at grails.transaction.GrailsTransactionTemplate.executeAndRollback(GrailsTransactionTemplate.groovy:64)
    at baseball.test.BiDirectionalAssociationSpecSpec.$tt__$spock_feature_0_0(BiDirectionalAssociationSpecSpec.groovy:20)
    at baseball.test.BiDirectionalAssociationSpecSpec.a valid team is bootstrapped into the test_closure2(BiDirectionalAssociationSpecSpec.groovy)
    at groovy.lang.Closure.call(Closure.java:426)
    at groovy.lang.Closure.call(Closure.java:442)
    at grails.transaction.GrailsTransactionTemplate$1.doInTransaction(GrailsTransactionTemplate.groovy:67)
    at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
    at grails.transaction.GrailsTransactionTemplate.executeAndRollback(GrailsTransactionTemplate.groovy:64)
    at baseball.test.BiDirectionalAssociationSpecSpec.a valid team is bootstrapped into the test(BiDirectionalAssociationSpecSpec
```

**Comment by jd73** [comment]
I am experiencing a same issue with grails 3.0.7

**Comment by denis111** [comment]
It happens in 2.5.1 also if there's beforeSave method in the domain class.

**Comment by graemerocher** [comment]
The issue is that the `player` property is unassigned when `save()` is called:

The following corrects the problem:
{code}
        def c = new Contract(
                expiration: new Date(),
                salary: 40_000_000
        )
        def p = new Player(
                firstName: "John",
                lastName: "Doe",
                position: "Pitcher",
                contract: c
        )
        c.player = p
        padres.addToPlayers(p)
{code}

Probably earlier versions of Grails aligned these associations before saving, I'll try track down the change, but it is not as serious as I thought and the above is a workaround

**Comment by joemccall86** [comment]
Thanks for taking a look at it. I'd like to help any way I could if you could point me in the right direction. A lot of our code was written based on that cascading behaviors, and my time would likely be better spent contributing a patch than adding this workaround everywhere we save an object.

**Comment by graemerocher** [comment]
I believe the problem is that in the latest versions of the Hibernate plugin this method is no longer present:

https://github.com/grails/grails-data-mapping/blob/3.x/grails-datastore-gorm-hibernate4/src/main/groovy/org/codehaus/groovy/grails/orm/hibernate/metaclass/AbstractSavePersistentMethod.java#L246

This is because the logic for `save()` was replaced / moved to:

https://github.com/grails/grails-data-mapping/blob/4.x/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/GormInstanceApi.groovy#L169

But obviously some behaviour seems to have not been retained. Since this is a Hibernate specific problem, it would probably make sense to override and provide an impl in the child class:

https://github.com/grails/grails-data-mapping/blob/4.x/grails-datastore-gorm-hibernate4/src/main/groovy/org/grails/orm/hibernate/HibernateGormInstanceApi.groovy

Thanks for the willingness to contribute.

**Comment by joemccall86** [comment]
Wow, it turns out I am in completely over my head on this one :-). 

Basically I started out trying to make the sample application into a unit test on the `grails-datastore-gorm-hibernate4` module, based on `BidirectionalOneToOneBelongsToCascadeSpec.groovy`, making it pass in the 3.x branch (the plugin version in grails 2.x), and seeing if it fails on the 4.x branch (the plugin used in grails 3.x). I couldn't even set up a passing test for the 3.x branch. The integration test with team->player->contract passes, but as soon as I replicate that in the unit test I get the same NPE failure as above (also that workaround makes it pass).

In other words, I can't seem to wrap my head around why Grails 2.x even works in this case. All I can contribute is this discovery: this bug doesn't only happen with hasMany. It can be reproduced with nested hasOne/belongsTo associations.

```
class Foo {
    static hasOne = [bar: Bar]
}

class Bar {
    static belongsTo = [foo: Foo]
    static hasOne = [baz: Baz]
}

class Baz {
    static belongsTo = [bar: Bar]
}

new Foo(
    bar: new Bar(
        baz: new Baz()
    )
).save(flush: true)
```

When I run the above code:

Project | Test Type | Runs
--------------|-----------|-----
Grails 2.4.5 | Integration | YES, all three rows inserted
Grails 3.0.8 | Integration | NO, NullPointerException above
grails-data-mapping 3.x branch | Unit | NO, NullPointerException above
grails-data-mapping 4.x branch | Unit | NO, NullPointerException above

I hope this helps in some small way. I need to move onto another task, but feel free to ask me anything about the tests I wrote, or any other questions.

**Comment by felansu** [comment]
Hi!

I implemented spring-security-core:3.0.0.M1 in my project and the same error !

Is reproducible in my project, BootStrap:17 when try save a user account !

**Comment by joemccall86** [comment]
I upgraded the test project to use 3.0.13 and added one for 3.1.0. Unfortunately the test still fails in both.

**Comment by graemerocher** [comment]
So I have added a test for this, this is not easy to fix at a GORM level because the only way to fix it is to modify the constructors via AST for domain classes so that during assignment the associations are aligned. Another option is to use AST to modify the setters to auto align the bidirectional one-to-one, again not a simple fix.

A third option that is less invasive and probably how it worked before is to modify the Grails data binding to align the properties. This would work only for when using GORM within a Grails project at the data binding constructor is only added via AST when you use Grails so would not work for example in standalone GORM.

I am undecided what the best fix is or whether this is even worth fixing yet since there is a workaround available and I do not want to add necessarily logic that traverses the whole relationship tree in order to automatically align associations as this would hurt performance. So the only potential future solution right now is an AST transformation that modifies the setters, but it is a non-trivial solution.

**Comment by magx2** [comment]
Is any update on this?

**Comment by davydotcom** [comment]
shouldnt the addToPlayer call make the reverse association for you?

**Comment by graemerocher** [comment]
@davydotcom The issue is not the `padres-players` association it is the `Contract` -> `player` association

**Comment by graemerocher** [comment]
Duplicates  #9820

**Comment by davydotcom** [comment]
My issue number is smaller though! 
:)

**Comment by graemerocher** [comment]
Very true!

**Comment by moskiteau** [comment]
What's the workaround? This?

`def oldRemoveBatchLoadableEntityKey = BatchFetchQueue.metaClass.&removeBatchLoadableEntityKey

        BatchFetchQueue.metaClass.removeBatchLoadableEntityKey = { EntityKey key ->
            return key ? oldRemoveBatchLoadableEntityKey.invoke(key) : null
        }`

**Comment by graemerocher** [comment]
The issue normally results from invalid application code. Either

a) Your domain model is not associated correctly prior to calling save
or
b) You are associating an invalid object with a valid one and trying to save said valid one

So the "workaround" is normally to fix your application logic.

**Comment by moskiteau** [comment]
![image](https://cloud.githubusercontent.com/assets/191202/18994553/5e7b030e-86f6-11e6-93f5-86e7c18a2862.png)

**Comment by graemerocher** [comment]
In Hibernate 5 the error message has been improved it seems. It is now:

```
Caused by: org.h2.jdbc.JdbcSQLException: NULL not allowed for column "PLAYER_ID"; SQL statement:
insert into contract (id, version, player_id, salary) values (null, ?, ?, ?) [23502-164]
```

**Comment by magx2** [comment]
Any update on this? Grails 3.2.4 still has the same problem and it's quite annoying.

**Comment by RichardKnowles** [comment]
I'm porting an ancient app from 1.3.6 to 2.5.6 and I've discovered exactly this case.

Class A in a bi-directional relationship with Class B.

My original code made the classic newbie Hibernate mistake where I have set up a reference from object A to object B, but no "back reference" from object B to object A.  Exactly as @graemerocher describes earlier in this issue (simplified here):

class A {
	static hasMany = [bInstances: B]
}

class B {
	static belongsTo = [parent: A]
}

Hibernate has always required for bi-directional relationships, both sides of the association are set.

a.bInstances << bInstance
bInstance.parent = a

So as @graemerocher pointed out, it was my invalid application code - the annoying thing was that the older version of Grails was erroneously kind of "autocorrecting" the error. It's not a workaround!
